### DISCUSSION TITLE
some close operation

### DISCUSSION THREAD

**Discussion by JaesukLim** [ANLAB-KAIST/KENSv3#194] [Project Q&A 2. Transport Layer (TCP)] [answered]
in our test cases, there is some cases that doesn't fit with the document's state transition diagram.
For example, for the BeforeAccept, at the close stage, the welcome socket sends ACK for the client's FIN/ACK packet, and the server tries to the connection socket. In this case, how does ack num and seq num can be shared?

Also I am wondering if I passed four test binary (kens-test-open, bind, handshake, close), the actual packet order and transfer is also taken account for grading?

**Comment by Pusnow** [comment]
> For example, for the BeforeAccept, at the close stage, the welcome socket sends ACK for the client's FIN/ACK packet, and the server tries to the connection socket. In this case, how does ack num and seq num can be shared?

I'm not sure what the "welcome socket" means, but SEQ/ACK numbers are determined in three-way-handshaking, which is done before the closing stage as well as the server part (tries to the connection socket?).

> Also I am wondering if I passed four test binary (kens-test-open, bind, handshake, close), the actual packet order and transfer is also taken account for grading?

As long as you pass all test cases, the packet order does not matter.

**Reply by JaesukLim** [reply]
Oh there was some misleading words
the welcome socket -> listening socket
What I wanted to ask is 
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/143942994/17f0c58d-f670-4f21-8b39-c4616f95594f)
in this picture, (from the BeforeAccept solution) when 5th ACK packet is sent, there is no connection socket and only listening socket exists. But the 6th FIN/ACK socket is sent by connection socket (by accept()). the ack num and seq num of these two packets are same.
so did you mean that the ack num and seq num for the handshaking is maintained for that connection?

**Reply by Pusnow** [reply, marked as answer]
> when 5th ACK packet is sent, there is no connection socket and only listening socket exists.

No, only client calls `close()`, so the server in *passive close* state. In this state, the server can send data until the server calls `close()`.
